samples/SkiaJsSample/src/jsMain/kotlin/org/jetbrains/skiko/sample/js/App.kt

```diff
@@ -2,46 +2,127 @@ package org.jetbrains.skiko.sample.js
 
-fun main() {
-    onWasmReady {
-        val paint = Paint().apply {
-            color4f = Color4f(0f, 1f, 0f, 1.0f)
-            mode = PaintMode.FILL
-            isAntiAlias = true
+private fun Canvas.draw(circle: Circle, paint: Paint) {
+    drawCircle(circle.x, circle.y, circle.r, paint)
+}
+
+private fun Circle.move(s: Float, angle: Float) {
+    x += s * sin(angle)
+    y += s * cos(angle)
+}
+
+private enum class Position {
+    INSIDE,
+    TOUCHES_SOUTH,
+    TOUCHES_NORTH,
+    TOUCHES_WEST,
+    TOUCHES_EAST
+}
+
+private fun calculatePosition(circle: Circle, boundingWidth: Int, boundingHeight: Int): Position {
+    val southmost = circle.y + circle.r
+    val northmost = circle.y - circle.r
+    val westmost = circle.x - circle.r
+    val eastmost = circle.x + circle.r
+
+    return when {
+        southmost >= boundingHeight -> Position.TOUCHES_SOUTH
+        northmost <= 0 -> Position.TOUCHES_NORTH
+        eastmost >= boundingWidth -> Position.TOUCHES_EAST
+        westmost <= 0 -> Position.TOUCHES_WEST
+        else -> Position.INSIDE
+    }
+}
+
+private class BouncingBall(
+    val circle: Circle,
+    val width: Int,
+    val height: Int,
+    val velocity: Float,
+    var angle: Double
+) {
+    fun recalculate(dt: Float) {
+        val position = calculatePosition(circle, width, height)
+
+        when (position) {
+            Position.TOUCHES_SOUTH -> angle = PI - angle
+            Position.TOUCHES_EAST -> angle = -angle
+            Position.TOUCHES_WEST -> angle = -angle
+            Position.TOUCHES_NORTH -> angle = PI - angle
         }
 
-        val canvas = document.getElementById("c") as HTMLCanvasElement
-        canvas.getContext("webgl")
-
-        GetWebGLContext(canvas)
-
-        val grContext = DirectContext.makeGL()
-
-        val surface = Surface.makeFromBackendRenderTarget(
-            grContext,
-            BackendRenderTarget.makeGL(
-                canvas.width, canvas.height, 1, 8, 0, 0x8058
-            ),
-            SurfaceOrigin.BOTTOM_LEFT,
-            SurfaceColorFormat.RGBA_8888,
-            ColorSpace.sRGB
-        )
-
-        fun draw() {
-            surface.canvas.drawCircle(50f, 50f, 25f, paint)
-            paint.color4f = Color4f(1f, 1f, 0f, 1.0f)
-            surface.canvas.drawCircle(100f, 100f, 10f, paint)
-            surface.canvas.drawCircle(200f, 200f, 30f, paint)
-            window.requestAnimationFrame { draw() }
-            surface.flushAndSubmit()
+        circle.move(velocity * (dt / 1000), angle.toFloat())
+    }
+}
+
+private fun Canvas.draw(ball: BouncingBall, paint: Paint) {
+    draw(ball.circle, paint)
+}
+
+private class BouncingBallsApp(htmlCanvas: HTMLCanvasElement, private var prevTimestamp: Double) : CanvasRenderer(htmlCanvas) {
+    private fun Color4f.asPaint() = Paint().apply {
+        color4f = this@asPaint
+        mode = PaintMode.FILL
+        isAntiAlias = true
+    }
+
+    val data = listOf(
```

Comment:
After switching tabs in browser rendering breaks, maybe canvas size is miscomputed or GL context gone.

Reply:
It looks like this is something that browser does intentionally  to save some energy from not running the process. This needs to be addressed but this particular PR neither introduced, nor, unfrotunately fixed this issue (

Reply:
Fixed, was not entirely correct bouncing logic for large deltas.